xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/ConversionMatrix.h

```diff
@@ -8,66 +8,102 @@
+#include <cmath>
 #include <memory>
 
 extern "C" {
 #include <libavutil/pixfmt.h>
 }
 
-template <unsigned Order>
+template<uint8_t Order>
 class CMatrix
 {
 public:
-  CMatrix(float (&src)[Order][Order]);
-  CMatrix(float (&src)[Order-1][Order-1]);
-  CMatrix& operator=(const CMatrix& src);
-  CMatrix& operator=(const float (&src)[Order-1][Order-1]);
+  CMatrix() = default;
+  explicit CMatrix(const CMatrix<Order - 1>& other);
+  explicit CMatrix(const std::array<std::array<float, Order>, Order>& other) { m_mat = other; }
+  explicit CMatrix(const std::array<std::array<float, Order - 1>, Order - 1>& other);
   virtual ~CMatrix() = default;
 
-  float (&Get())[Order][Order];
-  CMatrix Invert();
+  virtual CMatrix operator*(const std::array<std::array<float, Order>, Order>& other);
+
   CMatrix operator*(const CMatrix& other);
-  CMatrix operator*=(const CMatrix& other);
-  virtual CMatrix operator*(const float (&other)[Order][Order]);
+  CMatrix& operator*=(const CMatrix& other);
 
-protected:
-  CMatrix() = default;
+  CMatrix& operator=(const std::array<std::array<float, Order - 1>, Order - 1>& other);
+
+  bool operator==(const CMatrix<Order>& other) const
+  {
+    for (int i = 0; i < Order; ++i)
+    {
+      for (int j = 0; j < Order; ++j)
+      {
+        if (m_mat[i][j] == other.m_mat[i][j])
+          continue;
+
+        // some floating point comparisions should be done by checking if the difference is within a tolerance
+        if (std::abs(m_mat[i][j] - other.m_mat[i][j]) <=
+            (std::max(std::abs(other.m_mat[i][j]), std::abs(m_mat[i][j])) * 1e-2))
+          continue;
+
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  std::array<float, Order>& operator[](int index) { return m_mat[index]; }
+
+  const std::array<float, Order>& operator[](int index) const { return m_mat[index]; }
+
+  std::array<std::array<float, Order>, Order>& Get();
 
-  void Copy(float (&dst)[Order][Order], const float (&src)[Order][Order]);
-  void Invert(float (&dst)[Order][Order], float (&src)[Order][Order]);
+  const std::array<std::array<float, Order>, Order>& Get() const;
 
-  float m_mat[Order][Order] = {};
+  CMatrix& Invert() const;
+
+  float* ToRaw() { return &m_mat[0][0]; }
+
+protected:
+  std::array<std::array<float, Order>, Order> Invert(
+      std::array<std::array<float, Order>, Order>& other) const;
+
+  std::array<std::array<float, Order>, Order> m_mat{{}};
 };
 
 class CGlMatrix : public CMatrix<4>
 {
 public:
   CGlMatrix() = default;
-  CGlMatrix(float (&src)[3][3]);
+  explicit CGlMatrix(const CMatrix<3>& other);
+  explicit CGlMatrix(const std::array<std::array<float, 3>, 3>& other);
   ~CGlMatrix() override = default;
-  CMatrix operator*(const float (&other)[4][4]) override;
+  CMatrix operator*(const std::array<std::array<float, 4>, 4>& other) override;
 };
 
 class CScale : public CGlMatrix
 {
 public:
-  CScale(float x, float y, float z);
+  explicit CScale(float x, float y, float z);
```

Comment:
why is this explicit?

Reply:
explicit only if constructor has a single param? (i believe)

Reply:
@lrusak most constructors have params, but that doesn't mean you have to make them explicit :thinking:  Since C++11, you can make both single param and multi param constructors `explicit`. But it makes most sense for single param, in my opinion. You use the `explicit` keyword to prevent automatic conversion that you don't want to happen (because it wouldn't make sense, be confusing, or mess up type deduction). So for example, if you have the function `void test(CScale scale)`, you couldn't call it like this: `test({1.0f, 2.0f, 3.0f)}`. You would have to write out the type: `test(CScale{1.0f, 2.0f, 3.0f})`. That's just one example of the general idea behind `explicit`. Having said that, I'm just not sure it is warranted here. There doesn't seem to be a great risk associated with the automatic conversion.

Reply:
is there any harm though? On cppreference on of the examples on the `explicit` page shows a constructor with multiple parameters https://en.cppreference.com/w/cpp/language/explicit

Reply:
As written in my comment, it makes it impossible to use some shortcuts that C++ would otherwise allow you to use (see also the `// error: copy-list-initialization does not consider B::B(int,int)` on the page you linked to) I would turn the table around. What benefit does it provide? For single-argument constructors, it totally makes sense to mark some of them `explicit` because it's easy to get an automatic conversion that you didn't want. For multi-argument constructors, I really struggle to find a practical example where you need this behavior. In this specific case, I'd argue that you actually **want** the automatic conversion from 3 floats to a `CScale` to happen.

Reply:
okie